src/extension.ts

```diff
@@ -1731,17 +1675,80 @@ export async function saveFile(doc: vscode.TextDocument, datasetProvider: Datase
             vscode.window.showErrorMessage(err.message + "\n" + err.stack);
         }
     }
+    // Get specific node based on label and parent tree (session / favorites)
+    let nodes: ZoweNode[];
+    let isFromFavorites: boolean;
+    if (!sesNode || sesNode.children.length === 0) {
+        // saving from favorites
+        nodes = utils.concatChildNodes(datasetProvider.mFavorites);
+        isFromFavorites = true;
+    } else {
+        // saving from session
+        nodes = utils.concatChildNodes([sesNode]);
+        isFromFavorites = false;
+    }
+    node = await nodes.find((zNode) => {
+        // dataset in Favorites
+        if (zNode.contextValue === DS_FAV_CONTEXT) {
+            return (zNode.label === `[${sesName}]: ${label}`);
+        // member in Favorites
+        } else if (zNode.contextValue === DS_MEMBER_CONTEXT && isFromFavorites) {
+            const zNodeDetails = getProfileAndDataSetName(zNode);
+            return (`${zNodeDetails.profileName}(${zNodeDetails.dataSetName})` === `[${sesName}]: ${label}`);
+        } else if (zNode.contextValue === DS_MEMBER_CONTEXT && !isFromFavorites) {
+            const zNodeDetails = getProfileAndDataSetName(zNode);
+            return (`${zNodeDetails.profileName}(${zNodeDetails.dataSetName})` === `${label}`);
+        } else if (zNode.contextValue === DS_DS_CONTEXT) {
+            return (zNode.label.trim() === label);
+        } else {
+            return false;
+        }
+    });
+
+    // define upload options
+    let uploadOptions: IUploadOptions;
+    if (node) {
+        uploadOptions = {
+            etag: node.getEtag(),
+            returnEtag: true
+        };
+    }
+
     try {
-        const response = await vscode.window.withProgress({
+        const uploadResponse = await vscode.window.withProgress({
             location: vscode.ProgressLocation.Notification,
             title: localize("saveFile.response.save.title", "Saving data set...")
         }, () => {
-            return zowe.Upload.pathToDataSet(documentSession, doc.fileName, label);  // TODO MISSED TESTING
+            return zowe.Upload.pathToDataSet(documentSession, doc.fileName, label, uploadOptions);  // TODO MISSED TESTING
         });
-        if (response.success) {
-            vscode.window.showInformationMessage(response.commandResponse);  // TODO MISSED TESTING
+        if (uploadResponse.success) {
+            vscode.window.showInformationMessage(uploadResponse.commandResponse);  // TODO MISSED TESTING
+            // set local etag with the new etag from the updated file on mainframe
+            node.setEtag(uploadResponse.apiResponse[0].etag);
+        } else if (!uploadResponse.success && uploadResponse.commandResponse.includes(localize("saveFile.error.ZosmfEtagMismatchError", "Rest API failure with HTTP(S) status 412"))) {
```

Comment:
This could fail if the CLI localized message is different than the extension's
For example if their corresponding french translation didn't match, we won't be able to detect that it failed because of the etag stuff.
No need to take action on it.

Reply:
I was thinking the same thing at some point, and was wondering if it would fail in that case. Then somehow it slipped my mind... I guess we just need to compare against a literal string, since I assume z/OSMF only returns English error messages. 
Thanks for pointing it out.